### ISSUE TITLE
Workflows continue before submitted step properly finishes.

### ISSUE THREAD

**Issue by greschd** [aiidateam/aiida_core#1391]
When running multiple daemon workers, it can happen that a "parent" workflow continues before the "child" that it submitted properly finishes.

Workflows which show this behaviour (tested with 4 daemon workers):

```
import string

from aiida.orm.data.int import Int
from aiida.work.workchain import WorkChain, ToContext

OUTPUT_KEYS = list(string.ascii_letters)

class Parent(WorkChain):
    @classmethod
    def define(cls, spec):
        super(Parent, cls).define(spec)

        spec.outline(cls.submit_child, cls.retrieve_child)

    def submit_child(self):
        return ToContext(child=self.submit(Child))

    def retrieve_child(self):
        out_keys = self.ctx.child.get_outputs_dict().keys()
        self.report(out_keys)
        assert all(key in out_keys for key in OUTPUT_KEYS)


class Child(WorkChain):
    @classmethod
    def define(cls, spec):
        super(Child, cls).define(spec)

        spec.outline(cls.do_run)

    def do_run(self):
        for i, k in enumerate(OUTPUT_KEYS):
            self.out(k, Int(i))
```

When inspecting the child later, it does have all the outputs, but that's not the case at the point where the parent picks it up and checks for the outputs.

**Comment by greschd** [comment]
If I had to take a wild guess, I'd say that the child process is put in the ``Finished`` state while the output hasn't properly serialized / propagated across the DB yet.

**Comment by sphuber** [comment]
Fixed in #1396
